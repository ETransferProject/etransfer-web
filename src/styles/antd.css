html {
  --ant-primary-color: #285eff;
  --ant-primary-color-hover: #3b6cff;
  --ant-primary-color-active: #1e52f0;
  --ant-success-color: #00b73e;
  --ant-error-color: #f53f3f;
  --ant-error-color-hover: #fc7373;
  --ant-warning-color-active: #cf3737;
  --ant-warning-color: #f90;
  --ant-info-color: #285eff;
}
