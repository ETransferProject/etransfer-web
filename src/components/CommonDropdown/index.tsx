import React, { useState } from 'react';
import { Dropdown, DropdownProps, MenuProps } from 'antd';
import DynamicArrow, { TDynamicArrowSize } from 'components/DynamicArrow';
import clsx from 'clsx';
import styles from './styles.module.scss';

export interface CommonDropdownProps extends DropdownProps {
  getContainer?: string;
  childrenClassName?: string;
  isBorder?: boolean;
  menu?: Omit<MenuProps, 'onClick'>;
  hideDownArrow?: boolean;
  suffixArrowSize?: TDynamicArrowSize;
  /** use 'handleMenuClick' instead of 'onClick' */
  handleMenuClick?: (...args: Parameters<Required<MenuProps>['onClick']>) => void;
}

export default function CommonDropdown({
  getContainer,
  isBorder = true,
  childrenClassName,
  handleMenuClick,
  children,
  suffixArrowSize,
  ...props
}: CommonDropdownProps) {
  const [isOpen, setIsOpen] = useState(false);
  return (
    <Dropdown
      trigger={['click']}
      {...props}
      getPopupContainer={() => document.getElementById(getContainer as string) as HTMLElement}
      menu={{
        ...props.menu,
        onClick: (params) => {
          setIsOpen(false);
          handleMenuClick?.(params);
        },
      }}
      onOpenChange={(open) => {
        setIsOpen(open);
      }}>
      <div
        className={clsx(
          'cursor-pointer',
          'flex-row-center',
          styles['children-container'],
          isBorder && styles['children-container-border'],
          childrenClassName,
        )}>
        {children}
        {!props.hideDownArrow && (
          <DynamicArrow
            isExpand={isOpen}
            className={styles['children-icon']}
            size={suffixArrowSize}
          />
        )}
      </div>
    </Dropdown>
  );
}
