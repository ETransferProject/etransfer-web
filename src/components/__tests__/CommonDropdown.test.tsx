import { render, screen, act } from '@testing-library/react';
import { vi, describe, expect, test, beforeEach } from 'vitest';
import CommonDropdown from '../CommonDropdown';
import { Dropdown } from 'antd';
import DynamicArrow from 'components/DynamicArrow';
import clsx from 'clsx';

// Mock dependencies
vi.mock('antd', async (importOriginal) => {
  const actual = await importOriginal<typeof import('antd')>();
  return {
    ...actual,
    Dropdown: vi.fn(({ children, ...props }) => (
      <div data-testid="mock-dropdown" {...props}>
        {children}
      </div>
    )),
    Menu: vi.fn(() => <div data-testid="mock-menu" />),
  };
});

vi.mock('components/DynamicArrow', () => ({
  default: vi.fn(({ isExpand }) => <div data-testid="mock-arrow">{isExpand ? '↑' : '↓'}</div>),
}));

vi.mock('clsx', () => ({
  default: vi.fn(() => 'mocked-class'),
}));

describe('CommonDropdown Component', () => {
  const defaultProps = {
    children: <button data-testid="trigger">Trigger</button>,
    handleMenuClick: vi.fn(),
    menu: {
      items: [{ key: '1', label: 'Item 1' }],
    },
  };
  const mockHandleClick = vi.fn();
  const mockGetContainer = 'test-container';

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('renders basic structure', () => {
    render(<CommonDropdown {...defaultProps} />);

    // Assert that the Dropdown component is rendered
    expect(screen.getByTestId('mock-dropdown')).toBeInTheDocument();
    expect(screen.getByTestId('trigger')).toBeInTheDocument();
    expect(screen.getByTestId('mock-arrow')).toHaveTextContent('↓');
  });

  test('renders with default props', () => {
    render(
      <CommonDropdown getContainer={mockGetContainer}>
        <div data-testid="trigger" />
      </CommonDropdown>,
    );

    // Get the getPopupContainer handler from the Dropdown component
    const menuGetPopupContainerHandler = (Dropdown as any).mock.lastCall[0].getPopupContainer;
    // Call the getPopupContainer handler
    menuGetPopupContainerHandler?.();

    // Assert that the Dropdown component is called with the correct props
    expect(Dropdown).toHaveBeenCalledWith(
      expect.objectContaining({
        trigger: ['click'],
        getPopupContainer: expect.any(Function),
      }),
      expect.anything(),
    );
  });

  test('handles menu click correctly', () => {
    render(
      <CommonDropdown getContainer={mockGetContainer} handleMenuClick={mockHandleClick}>
        <div data-testid="trigger" />
      </CommonDropdown>,
    );

    // Simulate menu click
    const menuClickHandler = (Dropdown as any).mock.calls[0][0].menu.onClick;
    // Call the menu click handler
    menuClickHandler({ key: 'test' });

    // Assert that the handleMenuClick function is called with the correct argument
    expect(mockHandleClick).toHaveBeenCalledWith({ key: 'test' });
  });

  test('toggles dropdown state', () => {
    render(
      <CommonDropdown getContainer={mockGetContainer}>
        <div data-testid="trigger" />
      </CommonDropdown>,
    );

    // Simulate onOpenChange click
    const onOpenChange = (Dropdown as any).mock.calls[0][0].onOpenChange;
    // Call the onOpenChange handler
    act(() => {
      onOpenChange(true);
    });

    // Assert that the DynamicArrow component is called with the correct props
    expect(DynamicArrow).toHaveBeenCalledWith(expect.objectContaining({ isExpand: true }), {});
  });

  test('applies correct class names', () => {
    render(<CommonDropdown {...defaultProps} isBorder={false} childrenClassName="custom-class" />);

    // Check if the class names are merged correctly
    expect(clsx).toHaveBeenCalledWith(
      'cursor-pointer',
      'flex-row-center',
      expect.stringContaining('children-container'),
      false, // isBorder condition
      'custom-class',
    );
  });

  test('hides arrow when hideDownArrow=true', () => {
    const { rerender } = render(<CommonDropdown {...defaultProps} />);

    // Assert that the arrow is visible
    expect(screen.queryByTestId('mock-arrow')).toBeInTheDocument();

    // Rerender with hideDownArrow=true
    rerender(<CommonDropdown {...defaultProps} hideDownArrow={true} />);

    // Assert that the arrow is hidden
    expect(screen.queryByTestId('mock-arrow')).not.toBeInTheDocument();
  });

  test('uses getPopupContainer correctly', () => {
    // Create a test container
    const container = document.createElement('div');
    container.id = 'test-container';
    document.body.appendChild(container);

    render(<CommonDropdown {...defaultProps} getContainer="test-container" />);

    // Assert that the Dropdown component is called with the correct props
    expect(Dropdown).toHaveBeenCalledWith(
      expect.objectContaining({
        getPopupContainer: expect.any(Function),
      }),
      expect.anything(),
    );
  });

  test('passes suffixArrowSize to DynamicArrow', () => {
    render(<CommonDropdown {...defaultProps} suffixArrowSize="Small" />);

    // Assert that the DynamicArrow component is called with the correct props
    expect(DynamicArrow).toHaveBeenCalledWith(
      expect.objectContaining({
        size: 'Small',
      }),
      expect.anything(),
    );
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<CommonDropdown {...defaultProps} />);

    // Check if the snapshot matches the rendered component
    expect(asFragment()).toMatchSnapshot();
  });
});
