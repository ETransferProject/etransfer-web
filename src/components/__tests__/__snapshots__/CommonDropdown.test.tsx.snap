// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommonDropdown Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="mock-dropdown"
    menu="[object Object]"
    trigger="click"
  >
    <div
      class="mocked-class"
    >
      <button
        data-testid="trigger"
      >
        Trigger
      </button>
      <div
        data-testid="mock-arrow"
      >
        â†“
      </div>
    </div>
  </div>
</DocumentFragment>
`;
